### ISSUE TITLE
ForwardDiff v0.2.2 compiles really slowly

### ISSUE THREAD

**Issue by rgiordan** [JuliaDiff/ForwardDiff.jl#137]
Our [Celeste package](https://github.com/jeff-regier/Celeste.jl) depends extensively on `ForwardDiff.jl` for our [unit tests](https://github.com/jeff-regier/Celeste.jl/blob/master/test/test_derivatives.jl).  When we upgrade to v0.2.2 of `ForwardDiff.jl` we see massive slowdowns in compile time, on the order of 10x longer per function differentiated.  The result is that `ForwardDiff.jl` is too slow to be usable for testing, which for us is a big loss.  Furthermore, we cannot upgrade `Optim.jl` without also upgrading `ForwardDiff.jl` and slamming our test performance.

Note that we're dependent on julia 0.5.   I see that ForwardDiff's own unit tests take about 3x longer in 0.5 than 0.4.

Any thoughts?  Is there any chance that this compile time will come back down soon?

**Comment by mlubin** [comment]
Is this a regression between Julia 0.4 and Julia 0.5 or between ForwardDiff 0.1 and ForwardDiff 0.2 or both?

**Comment by rgiordan** [comment]
Both, I think, though I'm principally reporting a difference between
ForwardDiff 0.1 and 0.2 on Julia 0.5.  We cannot easily run our package on
0.4, so I'm not able to easily test the difference between the two Julia
versions with the same ForwardDiff version.  However, I also see that your
current unit tests run about 3x more slowly on Julia 0.5 than on Julia 0.4.

**Comment by rgiordan** [comment]
@jeff-regier reminded me that on Celeste, the old version of ForwardDiff had about the same performance on Julia 0.4 as 0.5.

**Comment by jrevels** [comment]
I've definitely noticed this as well, especially for Hessians.

**Comment by jrevels** [comment]
Part of it is likely that fact that Julia 0.5's function types give the compiler a lot more info to play with, and I wonder if the rewrite of `NTuple` to `Tuple{Vararg{T,N}}` could have affected compiler performance as well?

**Comment by KristofferC** [comment]
Different functions are different types now so we likely need to compile a lot more now than before when stuff was compiled once for every possible function input. In other words, `Function` used to be concrete, now it is abstract.

**Comment by tkoolen** [comment]
I've experienced this as well. Compilation time seems to increase with chunk size. As workaround, I set the chunk size for a Hessian in one of my unit tests to 1. Even with this workaround, Hessian compilation time seems to have increased w.r.t. 0.1.8.

**Comment by KristofferC** [comment]
Have you done any profiling? Is it type inference that takes time?

**Comment by KristofferC** [comment]
What happens if we put `::ANY` as the type for the function argument to disable specialisation on the function being differentiated.

**Comment by rgiordan** [comment]
When I run ```@profile``` on a slow ForwardDiff command, it does spend a
lot of cycles in ```inference.jl```.  I take that to mean that it's type
inference that's taking most of the time, but maybe someone more expert
than me should look at it to be sure.

**Comment by jrevels** [comment]
It would also be worth seeing what happens if we go back to using `@generated` functions for the [`Partials` operations](https://github.com/JuliaDiff/ForwardDiff.jl/blob/master/src/partials.jl#L124). IIRC, someone pointed out a while ago that having  so many methods on defined on tuple types could force the compiler to do a lot of work.

**Comment by KristofferC** [comment]
Def worth looking into.

**Comment by jrevels** [comment]
Alright, I did a test to see if the removal of `@generated` functions in the `Partials` partials code could've been the culprit.

Using master (input size is 100)
```julia
# includes compile time
julia> @time ForwardDiff.hessian!(out, rosenbrock, x)
 17.557256 seconds (2.57 M allocations: 68.467 MB, 0.24% gc time)

# does not include compile time
julia> @time ForwardDiff.hessian!(out, rosenbrock, x)
  0.006293 seconds (37 allocations: 89.375 KB)
```

Using master, but implementing [these operations](https://github.com/JuliaDiff/ForwardDiff.jl/blob/master/src/partials.jl#L124) as `@generated` functions:
```julia
# includes compile time
julia> @time ForwardDiff.hessian!(out, rosenbrock, x)
 16.807157 seconds (987.07 k allocations: 40.227 MB, 0.21% gc time)

# does not include compile time
julia> @time ForwardDiff.hessian!(out, rosenbrock, x);
  0.011444 seconds (37 allocations: 89.375 KB)
```

So the latter is a tad easier on the compiler, but causes an almost 2x runtime performance regression.

**Comment by KristofferC** [comment]
How is that possible? They should generate the same compiled code?

**Comment by jrevels** [comment]
> How is that possible? They should generate the same compiled code?

AFAIK, `@generated` functions limit the kind of optimizations the compiler can perform, which is one of the reasons why they're considered undesirable. 

@vtjnash might be able to lend some insight here.

**Comment by jrevels** [comment]
Ah, I had forgotten to force inlining in the `@generated` version. Here's the new timings for the `@generated` version:

```julia
julia> @time ForwardDiff.hessian!(out, rosenbrock, x);
 17.996713 seconds (2.65 M allocations: 71.953 MB, 0.27% gc time)

julia> @time ForwardDiff.hessian!(out, rosenbrock, x);
  0.006799 seconds (37 allocations: 89.375 KB)
```

So this version has slightly worse compile time performance, and about the same runtime performance. I put up a [gist](https://gist.github.com/jrevels/ec3c7474a41b5c166b13633e357f0891) with the `@generated` functions I used, if anybody's curious.

**Comment by KristofferC** [comment]
I tried with SnoopCompile and it went from like 15 seconds to 14 seconds... Pretty worthless.

**Comment by jrevels** [comment]
This seems to have been mostly fixed by #155, ref [my comment here](https://github.com/JuliaDiff/ForwardDiff.jl/pull/160#issuecomment-263658236).
